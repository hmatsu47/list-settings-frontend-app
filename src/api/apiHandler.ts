import { fetchWithTimeout } from "./fetchWithTimeout";

// ベース URI
export const baseUri = "/api";
export const baseUriTag = "/api";
export const remoteBaseUri = localStorage.getItem("remoteBaseUri");
export const remoteBaseUriTag = localStorage.getItem("remoteBaseUriTag");

// API からデータ取得（GET）
export const getApiData = async (url: string) => {
  return await (await fetchWithTimeout("GET", url)).json();
};
