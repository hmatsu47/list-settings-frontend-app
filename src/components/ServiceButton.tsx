import Button from "@suid/material/Button";
import { service, setService } from "../signal";

type Props = {
  serviceName: string;
};

export const ServiceButton = (props: Props) => {
  return (
    <Button
      variant={service() === props.serviceName ? "contained" : "outlined"}
      size="small"
      color={service() === props.serviceName ? "primary" : "inherit"}
      onClick={(e) => {
        setService(props.serviceName);
        // 選択肢をローカルストレージに記録しておく
        localStorage.setItem("selectedService", props.serviceName);
      }}
      sx={{ textTransform: "none" }}
      title={`${props.serviceName} を選択`}
    >
      {props.serviceName}
    </Button>
  );
};
