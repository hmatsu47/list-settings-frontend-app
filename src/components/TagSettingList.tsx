import { For, Show } from "solid-js";
import Box from "@suid/material/Box";
import Button from "@suid/material/Button";
import Paper from "@suid/material/Paper";
import Stack from "@suid/material/Stack";
import Table from "@suid/material/Table";
import TableBody from "@suid/material/TableBody";
import TableCell from "@suid/material/TableCell";
import TableContainer from "@suid/material/TableContainer";
import TableHead from "@suid/material/TableHead";
import TableRow from "@suid/material/TableRow";
import Typography from "@suid/material/Typography";
import { openUriInNewTab } from "../openUri";
import { tagSettings } from "../signal";

const convertEnvironmentName = (environmentName: string) => {
  switch (environmentName) {
    case "dev":
      return "開発";
    case "prod":
      return "本番";
  }
  return environmentName;
};

const buttonColor = (environmentName: string) => {
  const colors = localStorage.getItem("tagButtonColor");
  if (!colors) {
    return "#616161";
  }
  const map = new Map<string, string>(Object.entries(JSON.parse(colors)));
  return map.get(environmentName) ?? "#616161";
};

export const TagSettingList = () => {
  const headerTitle = localStorage.getItem("tagSettingsHeaderTitle");
  const uriPrefix = localStorage.getItem("tagSettingUriPrefix");
  const uriSuffix = localStorage.getItem("tagSettingUriSuffix");

  return (
    <Show when={tagSettings() !== undefined} fallback={<></>}>
      <Box
        sx={{
          width: "100%",
          minWidth: "1024px",
          display: "flex",
        }}
        aria-live="polite"
      >
        <Stack spacing={2} direction="column">
          <Typography variant="h5">{headerTitle}</Typography>
          <Show
            when={tagSettings() !== null}
            fallback={
              <Typography variant="h6">
                リポジトリにイメージがありません
              </Typography>
            }
          >
            <TableContainer component={Paper}>
              <Table aria-label="simple table">
                <TableHead>
                  <TableRow>
                    <TableCell>設定</TableCell>
                    <TableCell>リリース対象イメージタグ</TableCell>
                  </TableRow>
                </TableHead>
                <TableBody>
                  <For each={tagSettings()} fallback={<></>}>
                    {(settingItem) => (
                      <TableRow>
                        <TableCell component="th" scope="row">
                          <Button
                            variant="contained"
                            size="small"
                            // color="primary"
                            onClick={() => {
                              openUriInNewTab(
                                `${uriPrefix}${settingItem.environment_name}${uriSuffix}`
                              );
                            }}
                            sx={{
                              textTransform: "none",
                              backgroundColor: `${buttonColor(
                                settingItem.environment_name
                              )}`,
                            }}
                            title={`${convertEnvironmentName(
                              settingItem.environment_name
                            )}環境のリリース設定画面を開く`}
                          >
                            {convertEnvironmentName(
                              settingItem.environment_name
                            )}
                            環境設定
                          </Button>
                        </TableCell>
                        <TableCell>{settingItem.tags.join(", ")}</TableCell>
                      </TableRow>
                    )}
                  </For>
                </TableBody>
              </Table>
            </TableContainer>
          </Show>
        </Stack>
      </Box>
    </Show>
  );
};
