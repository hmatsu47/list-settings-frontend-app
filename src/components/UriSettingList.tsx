import { For, Show } from "solid-js";
import Box from "@suid/material/Box";
import Button from "@suid/material/Button";
import Paper from "@suid/material/Paper";
import Table from "@suid/material/Table";
import TableBody from "@suid/material/TableBody";
import TableCell from "@suid/material/TableCell";
import TableContainer from "@suid/material/TableContainer";
import TableHead from "@suid/material/TableHead";
import TableRow from "@suid/material/TableRow";
import Typography from "@suid/material/Typography";
import { service, uriRemoteSettings, uriSettings } from "../signal";
import { formatDateTimeDisplay } from "../formatDateTime";
import { UriSetting } from "../type";
import { openUriInNewTab } from "../openUri";

const convertEnvironmentName = (environmentName: string) => {
  switch (environmentName) {
    case "stg":
      return "ステージング";
    case "prod":
      return "本番";
  }
  return environmentName;
};

export const UriSettingList = () => {
  const headerTitle = localStorage.getItem("uriSettingsHeaderTitle");

  return (
    <Show
      when={
        (service() && uriRemoteSettings() !== undefined) ||
        uriSettings() !== undefined
      }
      fallback={<></>}
    >
      <Box
        sx={{
          width: "100%",
          minWidth: "1024px",
          display: "flex",
        }}
        aria-live="polite"
      >
        <Typography variant="h3">{headerTitle}</Typography>
        <Show
          when={uriRemoteSettings() !== null && uriSettings() !== null}
          fallback={
            <Typography variant="h6">
              リポジトリにイメージがありません
            </Typography>
          }
        >
          <TableContainer component={Paper}>
            <Table aria-label="simple table">
              <TableHead>
                <TableRow>
                  <TableCell>設定</TableCell>
                  <TableCell>イメージURI</TableCell>
                  <TableCell>リリース日時</TableCell>
                </TableRow>
              </TableHead>
              <TableBody>
                <SettingListParts settings={uriRemoteSettings()} />
                <SettingListParts settings={uriSettings()} />
              </TableBody>
            </Table>
          </TableContainer>
        </Show>
      </Box>
    </Show>
  );
};

type Props = {
  settings: UriSetting[] | null | undefined;
};

const SettingListParts = (props: Props) => {
  const uriPrefix = localStorage.getItem("uriSettingUriPrefix");
  const uriSuffix = localStorage.getItem("uriSettingUriSuffix");

  return (
    <Show when={props.settings} fallback={<></>}>
      <Box
        sx={{
          width: "100%",
          minWidth: "1024px",
          display: "flex",
        }}
        aria-live="polite"
      >
        <For each={props.settings} fallback={<></>}>
          {(settingItem) => (
            <Show
              when={settingItem.service_name === service()}
              fallback={<></>}
            >
              <TableRow>
                <TableCell component="th" scope="row">
                  <Button
                    variant="contained"
                    size="small"
                    color="primary"
                    onClick={() => {
                      openUriInNewTab(
                        `${uriPrefix}${settingItem.environment_name}${uriSuffix}`
                      );
                    }}
                    sx={{ textTransform: "none" }}
                    title={`${convertEnvironmentName(
                      settingItem.environment_name
                    )}環境のリリース設定画面を開く`}
                  >
                    {convertEnvironmentName(settingItem.environment_name)}
                    環境設定
                  </Button>
                </TableCell>
                <TableCell>
                  {settingItem.next_release
                    ? settingItem.next_release.image_uri
                    : settingItem.last_released.image_uri
                    ? settingItem.last_released.image_uri
                    : ""}
                </TableCell>
                <TableCell>
                  {settingItem.next_release
                    ? `次回：${formatDateTimeDisplay(
                        new Date(settingItem.next_release.release_at)
                      )}`
                    : settingItem.last_released
                    ? `前回：${formatDateTimeDisplay(
                        new Date(settingItem.last_released.released_at)
                      )}`
                    : ""}
                </TableCell>
              </TableRow>
            </Show>
          )}
        </For>
      </Box>
    </Show>
  );
};
