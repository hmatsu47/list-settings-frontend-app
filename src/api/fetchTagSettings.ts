import { ErrorResponse, TagSetting } from "../type";
import { baseUri, getApiData } from "./apiHandler";
import {
  message,
  messageSeverity,
  setMessage,
  setMessageSeverity,
  setTagSettings,
} from "../signal";

export const fetchTagSettings = async () => {
  const load = async (): Promise<void> => {
    const data: TagSetting[] | ErrorResponse = await getApiData(
      `${baseUri}/tagSettings`
    );
    if (
      typeof data === "object" &&
      data !== null &&
      typeof (data as ErrorResponse).message === "string"
    ) {
      // 戻り値がエラーメッセージの場合
      setTagSettings(undefined);
      // すでにエラーメッセージがセットされている場合は連結する（複数 API を続けて呼び出すため）
      const tempMessage =
        message() && messageSeverity() === "error" ? message() : "";

      setMessage(`${tempMessage}${(data as ErrorResponse).message}`);
      setMessageSeverity("error");
      return;
    }
    setTagSettings(data as TagSetting[]);
    // エラーメッセージは上書きしない
  };
  void load();
};
