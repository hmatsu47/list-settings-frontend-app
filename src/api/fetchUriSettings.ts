import { ErrorResponse, UriSetting } from "../type";
import { baseUri, remoteBaseUri, getApiData } from "./apiHandler";
import {
  message,
  messageSeverity,
  setMessage,
  setMessageSeverity,
  setRemoteServices,
  setServices,
  setUriRemoteSettings,
  setUriSettings,
} from "../signal";

// サービス名をユニーク化
const getUniqueServices = (data: UriSetting[]) => {
  const tempSetting = data.map((o) => {
    return o.service_name;
  });
  return [...new Set(tempSetting)];
};
// Local API or Remote API
const local = 1;

export const fetchImages = async (api: number) => {
  const load = async (): Promise<void> => {
    const data: UriSetting[] | ErrorResponse = await getApiData(
      `${api === local ? baseUri : remoteBaseUri}/uriSettings`
    );
    if (
      typeof data === "object" &&
      data !== null &&
      typeof (data as ErrorResponse).message === "string"
    ) {
      // 戻り値がエラーメッセージの場合
      if (api === local) {
        setServices(undefined);
        setUriSettings(undefined);
      } else {
        setRemoteServices(undefined);
        setUriRemoteSettings(undefined);
      }
      // すでにエラーメッセージがセットされている場合は連結する（複数 API を続けて呼び出すため）
      const tempMessage =
        message() && messageSeverity() === "error" ? message() : "";

      setMessage(`${tempMessage}${(data as ErrorResponse).message}`);
      setMessageSeverity("error");
      return;
    }
    if (api === local) {
      setUriSettings(data as UriSetting[]);
      setServices(getUniqueServices(data as UriSetting[]));
    } else {
      setUriRemoteSettings(data as UriSetting[]);
      setRemoteServices(getUniqueServices(data as UriSetting[]));
    }
    // エラーメッセージは上書きしない
  };
  void load();
};
