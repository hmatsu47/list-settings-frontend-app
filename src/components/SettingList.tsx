import { onMount } from "solid-js";
import Box from "@suid/material/Box";
import Stack from "@suid/material/Stack";
import {
  fetchLocal,
  fetchRemote,
  fetchUriSettings,
} from "../api/fetchUriSettings";
import { fetchTagSettings } from "../api/fetchTagSettings";
import { tagSettings, uriSettings, uriRemoteSettings } from "../signal";
import { TagSettingList } from "./TagSettingList";
import { UriSettingList } from "./UriSettingList";

export const SettingList = () => {
  onMount(() => {
    if (uriSettings() === undefined) {
      fetchUriSettings(fetchLocal);
    }
    if (uriRemoteSettings() === undefined) {
      fetchUriSettings(fetchRemote);
    }
    if (tagSettings() === undefined) {
      fetchTagSettings();
    }
  });

  return (
    <Stack spacing={2} direction="column">
      <UriSettingList />
      <Box
        sx={{
          padding: "10px 10px 0 10px",
          width: "100%",
          minWidth: "1024px",
          display: "flex",
          // justifyContent: "center",
        }}
        aria-live="polite"
      />
      <TagSettingList />
    </Stack>
  );
};
