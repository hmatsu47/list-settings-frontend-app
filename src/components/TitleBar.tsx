import AppBar from "@suid/material/AppBar";
import Box from "@suid/material/Box";
import IconButton from "@suid/material/IconButton";
import Stack from "@suid/material/Stack";
import Typography from "@suid/material/Typography";
import ReplayIcon from "@suid/icons-material/Replay";
import {
  fetchLocal,
  fetchRemote,
  fetchUriSettings,
} from "../api/fetchUriSettings";
import { fetchTagSettings } from "../api/fetchTagSettings";

const reload = () => {
  fetchUriSettings(fetchLocal);
  fetchUriSettings(fetchRemote);
  fetchTagSettings();
};

export const TitleBar = () => {
  return (
    <AppBar position="static">
      <Stack direction="row">
        <Typography
          variant="h6"
          component="div"
          sx={{
            flexGrow: 1,
            padding: "0 0 0 10px",
          }}
        >
          コンテナリリースイメージ一覧
        </Typography>
        <Box title="再読み込み">
          <IconButton
            size="small"
            edge="start"
            color="inherit"
            aria-label="reload"
            sx={{ mr: 1 }}
            onClick={() => reload()}
          >
            <ReplayIcon sx={{ fontSize: "large", padding: "2px" }} />
            Reload
          </IconButton>
        </Box>
      </Stack>
    </AppBar>
  );
};
