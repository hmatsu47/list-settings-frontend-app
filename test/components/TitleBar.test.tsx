// @vitest-environment jsdom

import { describe, expect, test, vi } from "vitest";
import "vi-fetch/setup";
import { mockFetch, mockGet } from "vi-fetch";
import { fireEvent, render } from "@solidjs/testing-library";
import { formatSnapshot } from "../common/formatSnapshot";
import { TitleBar } from "../../src/components/TitleBar";
import { baseUri, remoteBaseUri } from "../../src/api/apiHandler";
import { fetchTagSettings } from "../../src/api/fetchTagSettings";
import {
  fetchLocal,
  fetchRemote,
  fetchUriSettings,
} from "../../src/api/fetchUriSettings";
import {
  LastReleased,
  NextRelease,
  UriSetting,
  TagSetting,
} from "../../src/type";

describe("<TitleBar />", () => {
  beforeEach(() => {
    // localStorage はモック化すべきかも（うまくいかなかったため一旦モックなしで実装）
    localStorage.removeItem("baseUri");
    localStorage.removeItem("remoteBaseUri");
    mockFetch.clearAll();
  });
  afterEach(() => {
    vi.clearAllMocks();
    localStorage.removeItem("baseUri");
    localStorage.removeItem("remoteBaseUri");
  });
  const titleBarList = [
    {
      title: "タイトルバー表示",
      expected: "コンテナリリースイメージ一覧",
      baseUrl: "/api",
      remoteBaseUrl: "http://remote.example.com/api",
      uri_environment_name: "stg",
      last_released: {
        image_uri: "hogefuga",
        released_at: new Date(),
      } as LastReleased,
      next_release: {
        image_uri: "testtest",
        release_at: new Date(),
      } as NextRelease,
      service_name: "test1",
      remote_uri_environment_name: "prod",
      remote_last_released: {
        image_uri: "foobar",
        released_at: new Date(),
      } as LastReleased,
      remote_next_release: {
        image_uri: "testtesttest",
        release_at: new Date(),
      } as NextRelease,
      remote_service_name: "test2",
      tag_environment_name: "prod",
      tags: ["release", "testtest"],
    },
  ];
  titleBarList.forEach((testCase) => {
    test(testCase.title, async () => {
      localStorage.setItem("baseUri", testCase.baseUrl);
      localStorage.setItem("remoteBaseUri", testCase.remoteBaseUrl);
      const mockTag = mockGet(`${baseUri}/tagSettings`).willResolve([
        {
          environment_name: testCase.tag_environment_name,
          tags: testCase.tags,
        },
      ] as TagSetting[]);
      const mockUri = mockGet(`${baseUri}/uriSettings`).willResolve([
        {
          environment_name: testCase.uri_environment_name,
          last_released: testCase.last_released,
          next_release: testCase.next_release,
          service_name: testCase.service_name,
        },
      ] as UriSetting[]);
      const mockUriRemote = mockGet(`${remoteBaseUri}/uriSettings`).willResolve(
        [
          {
            environment_name: testCase.remote_uri_environment_name,
            last_released: testCase.remote_last_released,
            next_release: testCase.remote_next_release,
            service_name: testCase.remote_service_name,
          },
        ] as UriSetting[]
      );
      const { container, findByText, unmount } = render(() => <TitleBar />);
      const expected = (await findByText(testCase.expected)) as HTMLElement;
      expect(expected).toHaveTextContent(testCase.expected);
      const button = (await findByText("Reload")) as HTMLInputElement;
      expect(button).toHaveTextContent("Reload");
      // css の名前が動的に変わるので固定値に置換
      const htmlBefore = formatSnapshot(container.innerHTML);
      expect(htmlBefore).toMatchSnapshot();
      // Reload　ボタンクリック
      fireEvent.click(button);
      // とりあえず API の呼び出しが行われたことだけを確認（戻り値は今のところ上手くテストできず）
      expect(mockTag).toHaveFetched();
      expect(mockUri).toHaveFetched();
      expect(mockUriRemote).toHaveFetched();
      // css の名前が動的に変わるので固定値に置換
      const htmlAfter = formatSnapshot(container.innerHTML); // Before から変化しない
      expect(htmlAfter).toMatchSnapshot();
      unmount();
    });
  });
});
