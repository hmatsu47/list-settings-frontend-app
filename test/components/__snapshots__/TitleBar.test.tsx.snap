// Vitest Snapshot v1

exports[`<TitleBar /> > タイトルバー表示 1`] = `"<div class=\\"MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionStatic MuiAppBar-root MuiPaper-root css-xxxxxxxx\\"><div class=\\"MuiStack-root css-xxxxxxxx\\"><div class=\\"MuiTypography-root MuiTypography-h6 MuiTypography-root css-xxxxxxxx\\">コンテナリリースイメージ一覧</div><div title=\\"再読み込み\\" class=\\"MuiBox-root css-xxxxxxxx\\"><button type=\\"button\\" aria-label=\\"reload\\" tabindex=\\"0\\" class=\\"MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-edgeStart MuiIconButton-sizeSmall MuiIconButton-root MuiButtonBase-root css-xxxxxxxx\\"><svg aria-hidden=\\"true\\" viewBox=\\"0 0 24 24\\" focusable=\\"false\\" data-testid=\\"ReplayIcon\\" class=\\"MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSvgIcon-root css-xxxxxxxx\\"><path d=\\"M0 0h24v24H0z\\" fill=\\"none\\"></path><path d=\\"M12 5V1L7 6l5 5V7c3.31 0 6 2.69 6 6s-2.69 6-6 6-6-2.69-6-6H4c0 4.42 3.58 8 8 8s8-3.58 8-8-3.58-8-8-8z\\"></path></svg>Reload<span class=\\"MuiTouchRipple-root css-xxxxxxxx\\"></span></button></div></div></div>"`;

exports[`<TitleBar /> > タイトルバー表示 2`] = `"<div class=\\"MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionStatic MuiAppBar-root MuiPaper-root css-xxxxxxxx\\"><div class=\\"MuiStack-root css-xxxxxxxx\\"><div class=\\"MuiTypography-root MuiTypography-h6 MuiTypography-root css-xxxxxxxx\\">コンテナリリースイメージ一覧</div><div title=\\"再読み込み\\" class=\\"MuiBox-root css-xxxxxxxx\\"><button type=\\"button\\" aria-label=\\"reload\\" tabindex=\\"0\\" class=\\"MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-edgeStart MuiIconButton-sizeSmall MuiIconButton-root MuiButtonBase-root css-xxxxxxxx\\"><svg aria-hidden=\\"true\\" viewBox=\\"0 0 24 24\\" focusable=\\"false\\" data-testid=\\"ReplayIcon\\" class=\\"MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSvgIcon-root css-xxxxxxxx\\"><path d=\\"M0 0h24v24H0z\\" fill=\\"none\\"></path><path d=\\"M12 5V1L7 6l5 5V7c3.31 0 6 2.69 6 6s-2.69 6-6 6-6-2.69-6-6H4c0 4.42 3.58 8 8 8s8-3.58 8-8-3.58-8-8-8z\\"></path></svg>Reload<span class=\\"MuiTouchRipple-root css-xxxxxxxx\\"></span></button></div></div></div>"`;
